libbeat/statestore/backend/memlog/diskstore.go

```diff
@@ -165,5 +264,410 @@ func (s *diskstore) LogOperation(op op) error {
 // NOTE: due to limitation on some Operating system or file systems, the active
 //       marker is not a symlink, but an actual file.
 func (s *diskstore) WriteCheckpoint(state map[string]entry) error {
-	panic("TODO: implement me")
+	tmpPath, err := s.checkpointTmpFile(filepath.Join(s.home, "checkpoint"), state)
+	if err != nil {
+		return err
+	}
+
+	fileTxID := s.txid + 1
+	fileName := fmt.Sprintf("%v.json", fileTxID)
+	checkpointPath := filepath.Join(s.home, fileName)
+
+	if err := os.Rename(tmpPath, checkpointPath); err != nil {
+		return err
+	}
+	trySyncPath(s.home)
+
+	// clear transaction log once finished
+	s.checkpointClearLog()
+
+	// finish current on-disk transaction by increasing the txid
+	s.txid++
+
+	s.dataFiles = append(s.dataFiles, dataFileInfo{
+		path: checkpointPath,
+		txid: fileTxID,
+	})
+
+	// delete old transaction files
+	active, _ := activeDataFile(s.dataFiles)
+	updateActiveMarker(s.log, s.home, active)
+	s.removeOldDataFiles()
+
+	trySyncPath(s.home)
+	return nil
+}
+
+func (s *diskstore) checkpointTmpFile(baseName string, states map[string]entry) (string, error) {
+	tempfile := baseName + ".new"
+	f, err := os.OpenFile(tempfile, os.O_RDWR|os.O_CREATE|os.O_TRUNC|os.O_SYNC, s.fileMode)
+	if err != nil {
+		return "", err
+	}
+
+	ok := false
+	defer cleanup.IfNot(&ok, func() {
+		f.Close()
+	})
+
+	writer := bufio.NewWriterSize(&ensureWriter{f}, s.bufferSize)
+	enc := newJSONEncoder(writer)
+	if _, err = writer.Write([]byte{'['}); err != nil {
+		return "", err
+	}
+
+	first := true
+	for key, entry := range states {
+		prefix := []byte(",\n")
+		if first {
+			prefix = prefix[1:]
+			first = false
+		}
+		if _, err = writer.Write(prefix); err != nil {
+			return "", err
+		}
+
+		err = enc.Encode(storeEntry{
+			Key:    key,
+			Fields: entry.value,
+		})
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if _, err = writer.Write([]byte("\n]")); err != nil {
+		return "", err
+	}
+
+	if err = writer.Flush(); err != nil {
+		return "", err
+	}
+
+	if err = syncFile(f); err != nil {
+		return "", err
+	}
+
+	ok = true
+	if err = f.Close(); err != nil {
+		return "", err
+	}
+
+	return tempfile, nil
+}
+
+func (s *diskstore) checkpointClearLog() {
+	if s.logFile == nil {
+		s.logNeedsTruncate = true
+		return
+	}
+
+	err := s.logFile.Truncate(0)
+	if err == nil {
+		_, err = s.logFile.Seek(0, os.SEEK_SET)
+	}
+
+	if err != nil {
+		s.logFile.Close()
+		s.logFile = nil
+		s.logBuf = nil
+		s.logNeedsTruncate = true
+		s.logInvalid = true
+	}
+
+	s.logEntries = 0
+	s.logFileSize = 0
+}
+
+func updateActiveMarker(log *logp.Logger, home, active string) error {
```

Comment:
please document how `home` and `active` are used by this function

Reply:
renamed `home` to `homePath` and `active` to `checkpointFilePath` + add godoc.

```
// updateActiveMarker overwrites the active.dat file in the home directory with
// the path of the most recent checkpoint file.
// The active file will be written to `<homePath>`/active.dat.
func updateActiveMarker(log *logp.Logger, homePath, checkpointFilePath string) error {
```